x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,208 @@
+                }
+
+                final String address = getPublishedAddress(httpResponseHttp.getResponseBody());
+                if (address == null || address.isEmpty()) {
+                    throw new UserException(ExitCodes.CANT_CREATE,
+                        "Could not retrieve the ip address.");
+                }
+
+                final String stackVersion = getVersion(httpResponseHttp.getResponseBody());
+                if (stackVersion == null || stackVersion.isEmpty()) {
+                    throw new UserException(ExitCodes.CANT_CREATE,
+                        "Could not retrieve the version.");
+                }
+
+                final String fingerprint = SslUtil.calculateFingerprint(httpCaKeysAndCertificates.get(0).v2());
+
+                final XContentBuilder builder = JsonXContent.contentBuilder();
+                builder.startObject();
+                builder.startArray("adr");
+                builder.value(address);
+                builder.endArray();
+                builder.field("fgr", fingerprint);
+                builder.field("key", apiKey);//  CreateApiKeyResponse.getKey().toString());
+                builder.endObject();
+                final String jsonString = Strings.toString(builder);
+                final String token = Base64.getUrlEncoder().encodeToString(jsonString.getBytes(StandardCharsets.UTF_8));
+                return stackVersion + "." + token;
+            } catch (Exception e) {
+                logger.error(("Error generating enrollment token"), e);
+                throw new UserException(ExitCodes.CANT_CREATE, "Error generating enrollment token: " + e.getMessage());
+            }
+        }
+    }
+
+    private HttpResponse.HttpResponseBuilder responseBuilder(InputStream is, Terminal terminal) throws IOException {
+        final HttpResponse.HttpResponseBuilder httpResponseBuilder = new HttpResponse.HttpResponseBuilder();
+        if (is != null) {
+            byte[] bytes = toByteArray(is);
+            String responseBody = new String(bytes, StandardCharsets.UTF_8);
+            terminal.println(Terminal.Verbosity.VERBOSE, responseBody);
+            httpResponseBuilder.withResponseBody(responseBody);
+        } else {
+            terminal.println(Terminal.Verbosity.VERBOSE, "<Empty response>");
+        }
+        return httpResponseBuilder;
+    }
+
+    private byte[] toByteArray(InputStream is) throws IOException {
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        byte[] internalBuffer = new byte[1024];
+        int read = is.read(internalBuffer);
+        while (read != -1) {
+            baos.write(internalBuffer, 0, read);
+            read = is.read(internalBuffer);
+        }
+        return baos.toByteArray();
+    }
+
+    protected static URL createAPIKeyURL(URL url) throws MalformedURLException, URISyntaxException {
+        return new URL(url, (url.toURI().getPath() + "/_security/api_key").replaceAll("/+", "/"));
```

Comment:
why is the replaceAll needed here and below ?

Reply:
Not sure under which circumstances getPath can produce extra slashes; looked at how SetupPasswordTool is using it

Reply:
As I understand the default url may come from the publish_host settings and may or may not contain closing slash